src/mscorlib/src/System/Text/StringBuilder.cs

```diff
@@ -999,6 +999,124 @@ void ISerializable.GetObjectData(SerializationInfo info, StreamingContext contex
             return this;
         }
 
+        public StringBuilder Append(StringBuilder value)
+        {
+            Contract.Ensures(Contract.Result<StringBuilder>() != null);
+
+            if (value != null && value.Length > 0)
+            {
+                AppendCore(value, 0, value.Length);
+            }
+            return this;
+        }
+
+        public StringBuilder Append(StringBuilder value, int startIndex, int count)
+        {
+            if (startIndex < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(startIndex), Environment.GetResourceString("ArgumentOutOfRange_GenericPositive"));
+            }
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count), Environment.GetResourceString("ArgumentOutOfRange_GenericPositive"));
+            }
+            Contract.Ensures(Contract.Result<StringBuilder>() != null);
+            Contract.EndContractBlock();
+
+            if (value == null)
+            {
+                if (startIndex == 0 && count == 0)
+                {
+                    return this;
+                }
+                throw new ArgumentNullException(nameof(value));
+            }
+            if (count > value.Length - startIndex)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count), Environment.GetResourceString("ArgumentOutOfRange_IndexLength"));
+            }
+
+            if (count > 0)
+            {
+                AppendCore(value, startIndex, count);
+            }
+            return this;
+        }
+
+        private unsafe void AppendCore(StringBuilder value, int startIndex, int count)
```

Comment:
I think this will introduce a race condition between the copy and the caller of this method modifying the StringBuilder reference. This is the reason why in the framework we make defensive copies of strings before we mess with them.

One way to fix this is to not introduce this method at all and instead convert the `StringBuilder` into a string and pass that to the `Append` method. That sadly is not the best way though as it would hurt performance.

I think we need to figure out a way to solve this race condition and still maintain performace.

/cc @stephentoub @jkotas